### PR TITLE
Check for null in ReflectionTypeLoadException Message/ToString

### PR DESCRIPTION
**Pull request by wtgodbe**
For https://github.com/dotnet/coreclr/issues/15710

@jkotas PTAL

@sdmaclea how can I run the Arm64 tests that were failing previously?

### CODE REVIEW

**Comment by justinvp** [comment]
The `LoaderExceptions` array could also be null (the constructor does not validate that it is non-null). Should it be checked before it's used?

**Comment by wtgodbe** [comment]
@justinvp good call, updated the change to address.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs [hunk 1]
```diff
@@ -51,9 +51,15 @@ public override string Message
 
                 StringBuilder text = new StringBuilder();
                 text.AppendLine(base.Message);
-                foreach (Exception e in LoaderExceptions)
+                if (LoaderExceptions != null)
```
Comment (jkotas) [inline src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: The null check needs to happen before `LoaderException.Length == 0` above.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs [hunk 2]
```diff
@@ -51,9 +51,15 @@ public override string Message
...
                 {
-                    text.AppendLine(e.Message);
+                    foreach (Exception e in LoaderExceptions)
+                    {
+                        if (e != null)
+                        {
+                            text.AppendLine(e.Message);
```
Comment (jkotas) [inline src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: Not related to this change ... the AggregateException does not append these on new lines. Instead, it formats it as one long string. Is there a good reason why this is different?

(In any case, you should do some ad-hoc testing to verify that this works reasonably well.)

**Comment by jkotas** [comment]
@wtgodbe  Also, please add regression tests for the null argument and null exception to corefx.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs (hunk 2 above)
Comment (wtgodbe) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: I thought it would be more readable this way. It looks reasonable to me locally from the testing I've done.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs (hunk 1 above)
Comment (justinvp) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: Now that the null check has been added above, this one can be removed.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs (hunk 2 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: ok

**Comment by sdmaclea** [comment]
> @sdmaclea how can I run the Arm64 tests that were failing previously?

I ran it locally.  My guess is it is a PR1 test since it passed before it was merged.  I think Innerloop skips these.  @BruceForstall Do you know how to trigger PR1 tests

**Comment by BruceForstall** [comment]
My "hack" is to run one of the jitstress1 jobs, since those are guaranteed to be pri-1; there's some confusion about being able to run "basic" jobs pri-1.

e.g., one or more of:

```
Windows_NT arm Cross Checked jitstress1 Build and Test
Windows_NT arm64 Cross Checked jitstress1 Build and Test
Windows_NT armlb Cross Checked jitstress1 Build and Test
Windows_NT x64 Checked jitstress1
Windows_NT x64_arm64_altjit Checked jitstress1
Windows_NT x86 Checked jitstress1
Windows_NT x86_arm_altjit Checked jitstress1
```

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs [hunk 3]
```diff
@@ -62,11 +65,16 @@ public override string Message
         public override string ToString()
         {
             StringBuilder text = new StringBuilder();
-            text.AppendLine(SR.ReflectionTypeLoad_LoadFailed);
-
-            foreach (Exception e in LoaderExceptions)
+            text.AppendLine(base.ToString());
+            if (LoaderExceptions != null)
```
Comment (justinvp) [inline src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: Nit: It'd be nice if the check was moved to the top of the method (include the length check) and early exit to match the same flow as in `Message`.

A few other things:
 - It might be nice to avoid the extra NewLine that will be at the end of the resulting string.
 - Just a nit, but `LoaderExceptions` could be cached in a local.
 - The initial string can be passed to the `StringBuilder`'s constructor.

Altogether, something like:

```c#
        public override string Message
        {
            get
            {
                Exceptions[] exceptions = LoaderExceptions;
                if (exceptions == null || exceptions.Length == 0)
                {
                    return base.Message;
                }

                var text = new StringBuilder(base.Message);
                foreach (Exception e in exceptions)
                {
                    if (e != null)
                    {
                        text.AppendLine();
                        text.Append(e.Message);
                    }
                }
                return text.ToString();
            }
        }

        public override string ToString()
        {
            Exceptions[] exceptions = LoaderExceptions;
            if (exceptions == null || exceptions.Length == 0)
            {
                return base.ToString();
            }

            var text = new StringBuilder(base.ToString());
            foreach (Exception e in exceptions)
            {
                if (e != null)
                {
                    text.AppendLine();
                    text.Append(e.ToString());
                }
            }
            return text.ToString();
        }
```

**Comment by jkotas** [comment]
@wtgodbe Please sync and resolve the conflict

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs (hunk 3 above)
Comment (danmosemsft) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: This is not a hot path, I don't see the value of caching the field access.

**Comment by wtgodbe** [comment]
@dotnet-bot test Windows_NT x86 Checked Innerloop Build and Test

**Comment by justinvp** [comment]
The implementation of `Message` and `ToString` are so similar. How about sharing the implementation via a helper method?

Taking into account some of the feedback I mentioned [here](https://github.com/dotnet/coreclr/pull/15711#discussion_r159560557), something like:

```c#
public override string Message => CreateString(base.Message, e => e.Message);

public override string ToString() => CreateString(base.ToString(), e => e.ToString());

private string CreateString(string baseValue, Func<Exception, string> selector)
{
    if (LoaderExceptions == null || LoaderExceptions.Length == 0)
    {
        return baseValue;
    }

    var text = new StringBuilder(baseValue);
    foreach (Exception e in LoaderExceptions)
    {
        if (e != null)
        {
            text.AppendLine();
            text.Append(selector(e));
        }
    }
    return text.ToString();
}
```

**Comment by jkotas** [comment]
Need to merge to fix failing tests. Additional feedback can be submitted as separate PR.

File: src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs (hunk 3 above)
Comment (jkotas) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: > I don't see the value of caching the field access.

It tends to make the code smaller.

Comment (jkotas) [reply src/mscorlib/shared/System/Reflection/ReflectionTypeLoadException.cs]: Also, this is not a plain field access. It is a property access that is harder for JIT to optimize out.

E.g. if I change LoaderExceptions to be cached in a local in what the currently checked in implementation of Message property, the x64 native code will get `~6%` smaller (from 220 bytes to 206 bytes).
